### ISSUE TITLE
[Feature Request] Differentiable torch.histc and differentiable equalize

### ISSUE THREAD

**Issue by shijianjian** [kornia/kornia#790]
## 🚀 Feature
Our current implementation used a non-differentiable implementation of equalization. Mainly, it is caused by the non-differentiable ```torch.histc```. We may have different differentiable estimators to estimate the image histogram.

Presumably, I would expect an API as:
```python
def equalize(image, estimator=None):
    if estimator is None:
        hist = torch.histc(...)
    elif estimator == 'gaussian':
        hist = kornia.gaussian_hist(...)
    ...
```

## Additional context

There are some materials already to estimate the image histogram as [here](https://discuss.pytorch.org/t/differentiable-torch-histc/25865/27).

**Comment by bsuleymanov** [comment]
I've implemented differentiable image histogram using triangular kernel density estimation using bandwidth = (max - min) / n_bins. This histogram is approximate. How would you suggest to do test_uniform_dist or other tests for approximation?

**Comment by shijianjian** [comment]
> I've implemented differentiable image histogram using triangular kernel density estimation using bandwidth = (max - min) / n_bins. This histogram is approximate. How would you suggest to do test_uniform_dist or other tests for approximation?


Is it an implementation from a standard library? We may test the results against other libraries if appropriate. 

Or, you may simply test using a manual seed. Using Monte Carlo is even better but I have no idea how to do that. XD.

**Comment by bsuleymanov** [comment]
> Is it an implementation from a standard library? We may test the results against other libraries if appropriate.
> 
> Or, you may simply test using a manual seed. Using Monte Carlo is even better but I have no idea how to do that. XD.

No, it's not based on standard library.

I found that test_uniform_dist works for my implementation without changes :)

**Comment by bsuleymanov** [comment]
@shijianjian Could you give me an advice, what should I do to add differentiable image histogram (for (B, C, H, W)-shaped tensors) to Kornia? 

I thought equalize was not implemented yet, but now I have found it is. It works for (C, H, W) tensors now. Locally I get histograms for whole batch of images in vectorized way and then just iterate over batch and channels to equalize each the images. Is it useful to get equalization for the whole batch?

**Comment by shijianjian** [comment]
Can you provide a code snippet? If you have something already worked, you may go for a PR directly. We can iterate and optimize from there.

**Comment by bsuleymanov** [comment]
> Can you provide a code snippet? If you have something already worked, you may go for a PR directly. We can iterate and optimize from there.

`def image_hist2d(image: torch.Tensor, min: float = 0., max: float = 255.,
                 n_bins: int = 256, bandwidth: float = -1.,
                 centers: torch.Tensor = torch.tensor([]), return_pdf: bool = False):
    """Function that estimates the histogram of the input image(s).

    The calculation uses triangular kernel density estimation.

    Args:
        x: Input tensor to compute the histogram with shape
        :math:`(H, W)`, :math:`(C, H, W)` or :math:`(B, C, H, W)`.
        min: Lower end of the interval (inclusive).
        max: Upper end of the interval (inclusive). Ignored when
        :attr:`centers` is specified.
        n_bins: The number of histogram bins. Ignored when
        :attr:`centers` is specified.
        bandwidth: Smoothing factor. If not specified or equal to -1,
        bandwidth = (max - min) / n_bins.
        centers: Centers of the bins with shape :math:`(n_bins,)`.
        If not specified or empty, it is calculated as centers of
        equal width bins of [min, max] range.
        return_pdf: If True, also return probability densities for
        each bin.

    Returns:
        Computed histogram of shape :math:`(bins)`, :math:`(C, bins)`,
        :math:`(B, C, bins)`.
        Computed probability densities of shape :math:`(bins)`, :math:`(C, bins)`,
        :math:`(B, C, bins)`, if return_pdf is ``True``. Tensor of zeros with shape
        of the histogram otherwise.
    """
    if not isinstance(image, torch.Tensor):
        raise TypeError(f"Input image type is not a torch.Tensor. Got {type(image)}.")

    if centers is not None and not isinstance(centers, torch.Tensor):
        raise TypeError(f"Bins' centers type is not a torch.Tensor. Got {type(centers)}.")

    if centers.numel() > 0 and centers.dim() != 1:
        raise ValueError(f"Bins' centers must be a torch.Tensor "
                         "of the shape (n_bins,). Got {values.shape}.")

    if not isinstance(min, float):
        raise TypeError(f'Type of lower end of the range is not a float. Got {type(min)}.')

    if not isinstance(max, float):
        raise TypeError(f"Type of upper end of the range is not a float. Got {type(min)}.")

    if not isinstance(n_bins, int):
        raise TypeError(f"Type of number of bins is not an int. Got {type(n_bins)}.")

    if bandwidth != -1 and not isinstance(bandwidth, float):
        raise TypeError(f"Bandwidth type is not a float. Got {type(bandwidth)}.")

    if not isinstance(return_pdf, bool):
        raise TypeError(f"Return_pdf type is not a bool. Got {type(return_pdf)}.")

    device = image.device

    if image.dim() == 4:
        batch_size, n_channels, height, width = image.size()
    elif image.dim() == 3:
        batch_size = 1
        n_channels, height, width = image.size()
    elif image.dim() == 2:
        height, width = image.size()
        batch_size, n_channels = 1, 1
    else:
        raise ValueError(f"Input values must be a of the shape BxCxHxW, "
                         f"CxHxW or HxW. Got {image.shape}.")

    if bandwidth == -1.:
        bandwidth = (max - min) / n_bins
    if centers.numel() == 0:
        centers = min + bandwidth * (torch.arange(n_bins, device=device).float() + 0.5)
    centers = centers.reshape(-1, 1, 1, 1, 1)
    u = abs(image.unsqueeze(0) - centers) / bandwidth
    mask = (u <= 1).float()
    hist = torch.sum(((1 - u) * mask), dim=(-2, -1)).permute(1, 2, 0)
    if return_pdf:
        normalization = torch.sum(hist, dim=-1).unsqueeze(0) + 1e-10
        pdf = hist / normalization
        return hist, pdf
    return hist, torch.zeros_like(hist, dtype=hist.dtype, device=device)`
    
The histogram tests from kornia/tests are passed.

**Comment by shijianjian** [comment]
Thanks for the code. Some minor comments for the code here:
1. Probably we should have a generalised API and add a "kernel" parameter, which gonna be set to "triangular" at this stage. Also, if None, it gonna use ```torch.histc``` instead. We might implement more different estimators in the future.
2. Have you tested on CUDA? By reading the code, I think "centers" are not converted into CUDA device by default.

We always welcome vectorized implementations. This is definitely a good feature to add!

**Comment by bsuleymanov** [comment]
Thanks for the comments!

1. I agree with that. In this case I used triangular kde, because found this approach in [Learning Deep Embeddings with Histogram Loss](https://arxiv.org/abs/1611.00822) work.
2. Yes. When creating centers, we specify device `torch.arange(n_bins, device=device)`. I think we should test parametrize decorator to the perform tests on different devices.

**Comment by bsuleymanov** [comment]
I've add some changes (not in PR) to equalize and _scale_channel_functions to work with image_hist2d and different kernels, but do not know, what approximation error to accept in test. Do you have any ideas? Maybe MSE, MAE or MAPE w.r.t. equalized image using torch.histc.

**Comment by connorlee77** [comment]
Kornia's `histogram `function is already **differentiable** (via KDE) unless something changed. This appears to be more comprehensive in terms of selection of kernel, so consider replacing the original `histogram `function with this code.
